js/src/edit-free-campaign/getUnsavedShippingRates.js

```diff
@@ -18,15 +18,14 @@ import { isEqual, differenceWith, at } from 'lodash';
  * - Threshold: If `options` or `options.free_shipping_threshold` does not exist,
  *              `free_shipping_threshold` is treated as undefined.
  *
- * @param  {Array<ShippingRate>} rates The shipping rates to be compared.
- * @param  {Array<ShippingRate>} savedRates The saved shipping rates received from API.
- * @return {boolean} Whether there are any unsaved shipping changes.
+ * @param  {Array<ShippingRate>} newShippingRates The new shipping rates to be compared. This can be coming from Form values.
+ * @param  {Array<ShippingRate>} oldShippingRates The old shipping rates to be compared against. This should be previously received from API.
+ * @return {Array<ShippingRate>} Unsaved shipping rates from newShippingRates.
  */
-export default function hasUnsavedShippingRates( rates, savedRates ) {
-	if ( rates.length !== savedRates.length ) {
-		return true;
-	}
-
+export default function getUnsavedShippingRates(
```

Comment:
❓ 
The original `hasUnsavedShippingRates` also contains the logic to determine whether any deletion has occurred, but the `getUnsavedShippingRates` function only returns the changed shipping rates but does not include those deleted ones. Would it be better to use a name like `getDiffShippingRates` or something like that?

Further, instead of changing the original `hasUnsavedShippingRates` interface, I'd suggest extracting the following parts into a new function. When writing tests, `if ( rates.length ! == savedRates.length )` and `diffRates.length > 0` could be included in test cases together, rather than externalizing these logic to *edit-free-campaign/index.js*.

https://github.com/woocommerce/google-listings-and-ads/blob/34cacd305135b79ba0fda599a8f20db170404eaf/js/src/edit-free-campaign/hasUnsavedShippingRates.js#L30-L40

Reply:
> The original `hasUnsavedShippingRates` also contains the logic to determine whether _any deletion has occurred_, but the `getUnsavedShippingRates` function only returns the changed shipping rates but _does not include those deleted ones_.

🤦 My bad. When I saw the `if ( rates.length ! == savedRates.length )`, somehow I thought it is used to just short-circuit and quickly return a result without having to run `differenceWith` computation. I'll think of something to make things work.

Reply:
I have come up with a new function called `getDifferentShippingRates` in [`5f8649b` (#1313)](https://github.com/woocommerce/google-listings-and-ads/pull/1313/commits/5f8649b99da1ca81795ff1205857e02f8726e6b8). This should now be fixed.